packages/python/plotly/plotly/express/_core.py

```diff
@@ -1341,46 +1432,92 @@ def build_dataframe(args, constructor):
 
     # Cast data_frame argument to DataFrame (it could be a numpy array, dict etc.)
     df_provided = args["data_frame"] is not None
-    needs_interchanging = False
-    if df_provided and not isinstance(args["data_frame"], pd.DataFrame):
-        if hasattr(args["data_frame"], "__dataframe__") and version.parse(
-            pd.__version__
-        ) >= version.parse("2.0.2"):
-            import pandas.api.interchange
-
-            df_not_pandas = args["data_frame"]
-            args["data_frame"] = df_not_pandas.__dataframe__()
-            # According interchange protocol: `def column_names(self) -> Iterable[str]:`
-            # so this function can return for example a generator.
-            # The easiest way is to convert `columns` to `pandas.Index` so that the
-            # type is similar to the types in other code branches.
-            columns = pd.Index(args["data_frame"].column_names())
-            needs_interchanging = True
-        elif hasattr(args["data_frame"], "to_pandas"):
-            args["data_frame"] = args["data_frame"].to_pandas()
+    is_pd_like = False
+    if df_provided:
+
+        if nw.dependencies.is_polars_dataframe(
+            args["data_frame"]
+        ) or nw.dependencies.is_pyarrow_table(args["data_frame"]):
+            args["data_frame"] = nw.from_native(args["data_frame"], eager_only=True)
             columns = args["data_frame"].columns
-        elif hasattr(args["data_frame"], "toPandas"):
-            args["data_frame"] = args["data_frame"].toPandas()
+
+        elif nw.dependencies.is_polars_series(
+            args["data_frame"]
+        ) or nw.dependencies.is_pyarrow_chunked_array(args["data_frame"]):
+            args["data_frame"] = nw.from_native(
+                args["data_frame"],
+                series_only=True,
+            ).to_frame()
             columns = args["data_frame"].columns
-        elif hasattr(args["data_frame"], "to_pandas_df"):
-            args["data_frame"] = args["data_frame"].to_pandas_df()
+
+        elif nw.dependencies.is_pandas_like_dataframe(args["data_frame"]):
+
+            columns = args["data_frame"].columns  # This can be multi index
+            args["data_frame"] = nw.from_native(args["data_frame"])
+            is_pd_like = True
+
+        elif nw.dependencies.is_pandas_like_series(args["data_frame"]):
+
+            args["data_frame"] = nw.from_native(
+                args["data_frame"],
+                series_only=True,
+            ).to_frame()
             columns = args["data_frame"].columns
-        else:
-            args["data_frame"] = pd.DataFrame(args["data_frame"])
+            is_pd_like = True
+
+        elif hasattr(args["data_frame"], "__dataframe__"):
+            args["data_frame"] = nw.from_native(
+                nw.from_native(
+                    args["data_frame"], eager_or_interchange_only=True
+                ).to_pandas(),
+                eager_only=True,
+            )
```

Comment:
As before, interchange falls back to pandas, hence pandas is required for such case